code-styling/bash/README.md

```diff
@@ -0,0 +1,423 @@
+  it is sourced from a parent script. In other cases, `${0}` SHOULD be used.
+
+- As the `errexit` flag is used, `|| true` MUST be appended to commands that 
+  might cause an error which should not halt the script. (For instance when the
+  command output is assigned to a variable).
+
+### Command options
+
+- Long options MUST be used if they are available, 
+  for instance, use  `logger --priority --verbose` instead of `logger -p -v`. 
+
+### Command substitution
+
+- Sub-shell syntax `$(command)` MUST be used for command substitution rather
+  than backticks  <code>\`command`</code>.
+
+### Functions
+
+- Except for function packaging code (see below) all code MUST go in a function. 
+  Even if it's just one `main` function.
+
+- Functions MUST be defined without the `function` keyword, with round brackets, 
+  like this: `myfunc() { ... }`
+
+- All functions MUST be located together in the file just below constants. 
+  Executable code MUST NOT exists between functions.
+
+- Complex one-liners MUST be placed in standalone functions with an appropriate 
+  name and description.
+
+#### Function packaging
+
+- If it should be possible to call certain code both directly _and_ sourced from 
+  other scripts, it MUST be wrapped in a "self" check, like this: 
+  
+  ```bash
+  if [ "${BASH_SOURCE[0]}" == "${0}" ]; then
+      # direct call to file
+      main "${@}"
+  fi  # else file is included from another script
+  ```
+
+### Strings
+
+- For strings, single quotes `'` MUST be used when substitution is not required.
+- For strings, double quotes `"` MUST be used when substitution is required.
+
+### User messaging
+
+- User output SHOULD be formatted to follow the Heroku output format[^2](#references).
+- Warnings and errors MUST go to `STDERR`, anything than could be parsed SHOULD go to `STDOUT`.
+
+### Variables
+
+- Unless they are single character shell specials or arrays, variables MUST be 
+  "brace-quoted" (surrounded with curly braces `{` and `}`), 
+  like this: `${variable}`.  
+  Specials are `$!`, `$-`, `$_`, `$?`, `$#` `$*` `$@` `$$`.
+
+- Global variables SHOULD be avoided. When global variables _are_ used they MUST
+  be declared as `readonly` as much as possible.
+
+- Variables that are function-specific MUST be declared as `local`, unless `local` 
+  can not be used. (For instance when declaring as an associative array). For 
+  such cases `declare` MUST be used.
+ 
+- For the sake of clarity, read-only variables MUSt be declared using `readonly`.
+  Read-only variables MUST NOT be declared using `declare -r`.
+
+- Declaration and assignment SHOULD be placed on different lines.
+
+- To test variables that could be undeclared dot-slash `:-` MUST be used
+  (like this: `${variable:-}`)
+
+<!-- ======================================================================= -->
+
+## Common pitfalls
+
+Some more common problems and bugs in BASH can be avoided by adhering to the 
+following conventions:
+
```

Comment:
topic for the failures as
for i in  *.sql ; do .... ; done
find . -type f | xargs vs find . -type f -print0 | xargs -0
?